### ISSUE TITLE
Run workflow on demand

### ISSUE THREAD

**Issue by PeterMiklian** [AdhocAdam/smletsExchangeConnector#264]
I would consider useful to run SMLets Exchange Connector on demand in the GUI version.
I don't want to full the log unnecessarily every x seconds (current setting) while testing.
I would disable the workflow and click "Run now" only after saving changes in settings I want to test (and reopening settings or using task from SCSM console task pane?) to have only the related records in the log.
I don't know if this is possible, makes sense and what do you think about it :) Thank you for consideration, @AdhocAdam .

**Comment by AdhocAdam** [comment]
> I don't want to full the log unnecessarily every x seconds (current setting) while testing. I would disable the workflow and click "Run now" only after saving changes in settings I want to test (and reopening settings or using task from SCSM console task pane?) to have only the related records in the log.

Two follow up questions here:
1. If you're testing, why _wouldn't_ you want to see logs and operation as though it were actually running?
2. What changes are you making/how often are you making changes that merit this frequency of testing (i.e. on demand) vs. just setting a lower processing frequency while testing (e.g. 60 seconds) ?

**Comment by AdhocAdam** [comment]
> I don't know if this is possible

I think it is

> makes sense and what do you think about it

WELL! I've really been thinking on this a _lot_ and while I think it's probably possible, I don't think this is something that should be pursued for a few reasons. But rather than just a "no", let me walk you through my thought process on this:

1. Per my previous comment, you could just lower the workflow interval. This mimics the stock connector's functionality.
2. The stock connector:
    - has an interesting choice of verbiage that _surely_ must exist for a reason "Only use a polling interval less than 2 minutes in a demo/test configuration!"
        - ![stockSMExco](https://user-images.githubusercontent.com/6636040/109397740-da551480-78fd-11eb-9a81-00f1cec29b85.png)

    - is the only connector that _does not_ have a "Sync Now" task

Now from those points a few things come to mind:
- should a "Sync Now" to be introduced similiar to AD, SCO, or SCOM connectors. There isn't a way to prevent the workflow from running several times in parallel and creating duplicate Work Items. Certainly not at least without introducing more checks and balances. No less **explicitly** for this sole Task, that is worth pointing out only an Administrator could execute and otherwise _would not_ during production runs
- it could also be that since the stock connectors have a runtime longer than a couple minutes, it's easy to check their "Percent Complete" and then just say "don't run if not 100%" since they are part of the System Center connector framework. Which means you could potentially spam the Task several times, have several concurrent jobs, and your workflow server pays the price for it
- I have to imagine the reason all of the Microsoft SCSM Connectors (except Exchange) have a "Sync Now" option is because there isn't a concern of creating duplicate data with them. As all of the connectors create and update Config Items. Whereas the Exchange Connector is the _only one_ that creates and updates Work Items. The same Work Items people interact with, respond to, notify on, automate on, and report on. Whereas this risk is minimal if not non existant with Config Items and their connectors. It's not to say you can't or aren't doing automation based on changes to Config Items but if you are - you'd be creating Work Items for people to take action on. 

In regards to logging, if there is a need to test individual functionality out within the connector. I strongly reccomend trying it out in a dev SCSM environment so you can observe logging as it would occur in production. You could also take an Orchestrator approach of cranking up logging for testing, but lowering it once you're up and running. Plus in a similiar Orchestrator vein, you could just clear the logs and in this case use PowerShell's Clear-EventLog. Perhaps something with Custom Events?

It's with all of that said. I think even with extreme caution of making sure you disable the workflow and then run it leaves me uneasy. Because should this feature be pursued, it's creating an issue that someone could unknowingly walk into where none currently exists.

**Comment by PeterMiklian** [comment]
Adam, thank you for sharing your minds about this.
Here's situation which lead me to write this _Feature request_: I'm new to SCSM and want to implement your _SMLets Exchange connector_. I know what I want so I set it up, enable and ... something is not working as expected. Or want to see what happens after any tiny change in settings. As I want to have all data available for debugging I set log level to verbose and scheduled connector to run every 60 seconds. Meanwhile I have to study something, make infrastructure changes, modify Exchange settings, prepare testing work items/e-mails, go home after work, whatever :) I come back to testing and see log full of "initializing connector, encryption module enabled" etc. - records written to log every 60 seconds connector runs disregarding if the mailbox is empty or not. I have to clear the log or ignore all those items to see relevant events about the "real" run.
Yes, I can disable the connector each time, go back to settings and re-enable it after all changes I wanted to do - which is basically the same as having the "only run now" task while connector is disabled.

All those facts about other connectors are correct. A few more words:

1. sometimes AD (or other connector) get stuck on xy %. Clicking on "_Synchronize now_" should force it to stop and to run new instance as far as I remember.
2. _Asset Import connector_ has a brilliant setting "test mode only" (=logging)
3. checking "progress = 100 %" seems to be good indicator of the connector status. Running more parallel instances of _SMLets Exchange Connector_ makes no sense, sure.

**Comment by AdhocAdam** [comment]
>sometimes AD (or other connector) get stuck on xy %. Clicking on "Synchronize now" should force it to stop and to run new instance as far as I remember.

> checking "progress = 100 %" seems to be good indicator of the connector status. Running more parallel instances of SMLets Exchange Connector makes no sense, sure.

What I'm saying is that by definition this can't be done because the SMLets Exchange Connector is not part of the SCOM/SCSM Connector Framework. So there is no progress/percent complete that could be checked.

But I suppose my takeaway here is that it sounds like there could be something done to improve the nature of logging and logging levels.

**Comment by PeterMiklian** [comment]
You still can use old school simple "lock file exists" as an indicator of running instance :) Check existence of the file "connector_running.txt" on start, run connector and create it if it doesn't exist and delete it on finish to "release the lock" - not nice but functional.
Yes, I think I will have some more questions/suggestions about logging. I just need more time and looking into logs to sort out my minds :)

**Comment by AdhocAdam** [comment]
Closing as this will not be worked on. Logging will continue to receive updates and enhancements on subsequent releases of the connector.
